blockchain/blockchain.go

```diff
@@ -240,9 +247,213 @@ func NewBlockChain(db database.DBManager, cacheConfig *CacheConfig, chainConfig
 	// Take ownership of this particular state
 	go bc.update()
 	go bc.gcCachedNodeLoop()
+	go bc.checkRestartStateMigration()
+
 	return bc, nil
 }
 
+type stateTrieMigrationDB struct {
+	database.DBManager
+}
+
+func (td *stateTrieMigrationDB) ReadCachedTrieNode(hash common.Hash) ([]byte, error) {
+	return td.ReadCachedTrieNodeFromNew(hash)
+}
+func (td *stateTrieMigrationDB) ReadCachedTrieNodePreimage(secureKey []byte) ([]byte, error) {
+	return td.ReadCachedTrieNodePreimageFromNew(secureKey)
+}
+
+func (td *stateTrieMigrationDB) ReadStateTrieNode(key []byte) ([]byte, error) {
+	return td.ReadStateTrieNodeFromNew(key)
+}
+
+func (td *stateTrieMigrationDB) HasStateTrieNode(key []byte) (bool, error) {
+	return td.HasStateTrieNodeFromNew(key)
+}
+
+func (td *stateTrieMigrationDB) ReadPreimage(hash common.Hash) []byte {
+	return td.ReadPreimageFromNew(hash)
+}
+
+func (bc *BlockChain) migrateState(rootHash common.Hash) error {
+	bc.wg.Add(1)
+	defer bc.wg.Done()
+
+	srcCachedDB := bc.StateCache().TrieDB()
+	targetDB := statedb.NewDatabase(&stateTrieMigrationDB{bc.db})
+
+	trieSync := state.NewStateSync(rootHash, targetDB.DiskDB())
+	maxBatchSize := 10000
+	queue := append([]common.Hash{}, trieSync.Missing(maxBatchSize)...)
+
+	committedCnt := 0
+
+	for trieSync.Pending() > 0 {
+		// TODO-Klaytn refine the status parameter (progress percentage, etc)
+		bc.committedCnt, bc.pendingCnt = committedCnt, trieSync.Pending()
+		logger.Warn("State migration progress", "committedCnt", bc.committedCnt, "pendingCnt", bc.pendingCnt)
+
+		results := make([]statedb.SyncResult, len(queue))
+
+		for i, hash := range queue {
+			data, err := srcCachedDB.NodeFromOld(hash)
+			if err != nil {
+				return fmt.Errorf("failed to retrieve node data for %x: %v", hash, err)
+			}
+
+			results[i] = statedb.SyncResult{Hash: hash, Data: data}
+		}
+
+		committedCnt += len(results)
+
+		if _, index, err := trieSync.Process(results); err != nil {
+			return fmt.Errorf("failed to process result #%d: %v", index, err)
+		}
+
+		if index, err := trieSync.Commit(targetDB.DiskDB().GetStateTrieMigrationDB()); err != nil {
+			return fmt.Errorf("failed to commit data #%d: %v", index, err)
+		}
+
+		queue = append(queue[:0], trieSync.Missing(maxBatchSize)...)
+
+		select {
+		case <-bc.stopStateMigration:
+			// TODO-Klaytn Revert DB.
+			// - copied new DB data to old DB.
+			// - remove new DB
+			return errors.New("stop state migration")
+		case <-bc.quit:
+			return nil
+		default:
+		}
+	}
+	logger.Info("completed to copy state migration", "resultCnt", committedCnt)
+
+	// Preimage Copy
+	// TODO-Klaytn consider to copy preimage
+
+	// Cross check that the two tries are in sync
+	// TODO-Klaytn consider to check Trie contents optionally
+	// TODO-Klaytn consider to check storage trie also
+	dirty, err := bc.checkTrieContents(targetDB, srcCachedDB, rootHash)
```

Comment:
Why don't you check the state root after recalculating it?

Reply:
I think comparing all state tries is enough.
If there is crashed trie, it can be detected?

Reply:
I am afraid of whether `checkTrieContents` operates correctly as it is only used for test purposes.
Besides, calculating state root has been used to check integrity of states.
Anyway, `checkTrieContents` is also good to go if you fully tested.

Reply:
I am testing `checkTrieContents` works well. 
And I also consider to support checking stroage trie also. 
It will be introduced in next PR. :)

If this it not enough, I will add more routine to check as you mentioned.